### PR TITLE
[clang] Introduce `SemaSYCL`

### PR DESCRIPTION
**Pull request by Endilll**
This patch moves SYCL-related `Sema` functions into new `SemaSYCL` class, following the recent example of OpenACC and HLSL. This is a part of the effort to split `Sema`. Additional context can be found in #82217, #84184, #87634.

### CODE REVIEW

File: clang/include/clang/Sema/SemaSYCL.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
+//===----- SemaOpenACC.h 000- Semantic Analysis for SYCL constructs -------===//
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaSYCL.h:1]:
```suggestion
//===----- SemaSYCL.h 000- Semantic Analysis for SYCL constructs -------===//
```

File: clang/include/clang/Sema/SemaSYCL.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  ///   realize that the function will be codegen'ed.
+  ///
+  /// Example usage:
+  ///
+  /// Diagnose __float128 type usage only from SYCL device code if the current
+  /// target doesn't support it
+  /// if (!S.Context.getTargetInfo().hasFloat128Type() &&
+  ///     S.getLangOpts().SYCLIsDevice)
+  ///   SYCLDiagIfDeviceCode(Loc, diag::err_type_unsupported) << "__float128";
+  SemaDiagnosticBuilder SYCLDiagIfDeviceCode(SourceLocation Loc,
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaSYCL.h:47]:
```suggestion
  SemaDiagnosticBuilder DiagIfDeviceCode(SourceLocation Loc,
```

FWIW, removing the manual namespacing I found to be a very nice change for OpenACC, so I'd suggest we do it here too.

File: clang/include/clang/Sema/SemaSYCL.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                                             unsigned DiagID);
+
+  void deepTypeCheckForSYCLDevice(SourceLocation UsedAt,
+                                  llvm::DenseSet<QualType> Visited,
+                                  ValueDecl *DeclToCheck);
+
+  ExprResult BuildSYCLUniqueStableNameExpr(SourceLocation OpLoc,
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaSYCL.h:54]:
```suggestion
  ExprResult BuildUniqueStableNameExpr(SourceLocation OpLoc,
```

Perhaps?

File: clang/include/clang/Sema/SemaSYCL.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  /// Creates a SemaDiagnosticBuilder that emits the diagnostic if the current
+  /// context is "used as device code".
+  ///
+  /// - If CurLexicalContext is a kernel function or it is known that the
+  ///   function will be emitted for the device, emits the diagnostics
+  ///   immediately.
+  /// - If CurLexicalContext is a function and we are compiling
+  ///   for the device, but we don't know that this function will be codegen'ed
+  ///   for devive yet, creates a diagnostic which is emitted if and when we
```
Comment (hvdijk) [inline clang/include/clang/Sema/SemaSYCL.h:37]: This took me a moment longer to parse than it should. I realise you only moved it, this is not newly written by you, but maybe good to update at the same time by moving the "yet" to the front to make it clear and unambiguous that it applies to the "know", not to the "emitted":
```suggestion
  ///   for the device, but we don't know yet that this function will be codegen'ed
  ///   for the device, creates a diagnostic which is emitted if and when we
```

File: clang/include/clang/Sema/SemaSYCL.h (hunk 3 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaSYCL.h]: Done.

File: clang/include/clang/Sema/SemaSYCL.h (hunk 2 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaSYCL.h]: I did that for all 4 functions in `SemaSYCL`.

File: clang/include/clang/Sema/SemaSYCL.h (hunk 1 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaSYCL.h]: That's... embarrassing.

**Review by AaronBallman — APPROVED** [review]
LGTM

**Review by erichkeane — APPROVED** [review]

**Pull request merged**